src/c++/perf_analyzer/genai-perf/genai_perf/llm_inputs/synthetic_prompt_generator.py

```diff
@@ -67,26 +67,54 @@ def _create_farewell_lines(cls) -> List[str]:
     @classmethod
     def _create_prompt_from_farewell_lines(
         cls,
-        remaining_prompt_tokens: int,
+        requested_prompt_tokens: int,
         farewell_lines: List[str],
         tokenizer: Tokenizer,
     ) -> str:
-        prompt = ""
         get_token_length = lambda text: len(tokenizer.encode(text))
 
-        sampling_lines = True
-        while sampling_lines:
-            for line in farewell_lines:
-                line_to_add = line
-                if remaining_prompt_tokens - get_token_length(line_to_add) < 0:
-                    # This will cut off a line in the middle of a word, but that's ok since an
-                    # llm should be able to handle that.
-                    line_to_add = line_to_add[: int(math.ceil(remaining_prompt_tokens))]
-                    sampling_lines = False
-                    prompt += line_to_add
-                    break
-                prompt += line_to_add
-                remaining_prompt_tokens -= get_token_length(line_to_add)
+        line_iterator = itertools.cycle(farewell_lines)
+
+        def word_generator():
+            while True:
+                next_line = next(line_iterator)
+                words = re.split("[ \n]+", next_line)
+                for word in words:
+                    yield word
+
+        word_iterator = word_generator()
+
+        # FAST ADD lines
+        remaining_tokens = requested_prompt_tokens
+        prompt = ""
+        num_tokens_in_avg_line = (
+            get_token_length(farewell_lines[0] + farewell_lines[1]) / 2
+        )
+        num_lines_to_add_fast = math.floor(
+            0.5 * requested_prompt_tokens / num_tokens_in_avg_line
+        )
+        while num_lines_to_add_fast:
+            for _ in range(num_lines_to_add_fast):
+                next_line = next(line_iterator)
+                prompt = prompt + next_line
+
+            curr_tokens = get_token_length(prompt)
+            remaining_tokens = requested_prompt_tokens - curr_tokens
+            num_lines_to_add_fast = math.floor(
+                0.5 * remaining_tokens / num_tokens_in_avg_line
+            )
+
+        # FAST ADD words
+        final_line = ""
+        while get_token_length(final_line) < remaining_tokens - 3:
+            next_word = next(word_iterator)
+            final_line += next_word + " "
+        prompt += final_line
+
+        # Final tweaks
+        diff = requested_prompt_tokens - get_token_length(prompt)
+        for _ in range(diff):
+            prompt += "hi "
```

Comment:
in the case where we ask for 1024 tokens, how many will this generate?
i am assuming no guarantee here but what is the margin?
1024 +- 2 or 10 or 100?
We were way off before, using industry standard practices.
I am curious how much better this is.

Reply:
The above lines will get it within 2 or 3. This will add at most 3 "hi ".

Reply:
seriously?! 
wow.